File: 2022/09.py
from utils.grid import NEIGH_DICT


def get_input(puzzle: str) -> list[tuple[int, int, int]]:
    return [(*NEIGH_DICT[line[0]], int(line[2:])) for line in puzzle.splitlines()]


def solve(lines, n):
    knots: list[tuple[int, int]] = [(0, 0) for _ in range(n)]
    visited = {knots[-1]}
    for dx, dy, cnt in lines:
        for _ in range(cnt):
            knots[0] = knots[0][0] + dx, knots[0][1] + dy
            for i in range(1, n):
                p = knots[i - 1][0] - knots[i][0]
                q = knots[i - 1][1] - knots[i][1]
                if abs(p) <= 1 and abs(q) <= 1:
                    continue
                knots[i] = knots[i][0] + max(-1, min(p, 1)), knots[i][1] + max(-1, min(q, 1))
            visited.add(knots[-1])
    return len(visited)


def part1(puzzle: str):
    lines = get_input(puzzle)
    return solve(lines, 2)


def part2(puzzle: str):
    lines = get_input(puzzle)
    return solve(lines, 10)


if __name__ == "__main__":
    from aoc import run

    run(2022, 9, part1, part2)

File: 2022/09.rs
use aoc::grid::Direction;
use rustc_hash::FxHashSet;

type Input = Vec<Motion>;

#[derive(Debug)]
struct Motion {
    direction: Direction,
    count: u32,
}

struct Solver<'a> {
    motions: &'a [Motion],
    knots: Vec<(isize, isize)>,
    visited: FxHashSet<(isize, isize)>,
}

impl<'a> Solver<'a> {
    fn new(input: &'a Input, n: usize) -> Self {
        Self {
            motions: input,
            knots: vec![(0, 0); n],
            visited: FxHashSet::default(),
        }
    }

    fn follow(&mut self, i: usize) {
        let p = self.knots[i - 1].0 - self.knots[i].0;
        let q = self.knots[i - 1].1 - self.knots[i].1;
        if p.abs() > 1 || q.abs() > 1 {
            self.knots[i] = (
                self.knots[i].0 + p.max(-1).min(1),
                self.knots[i].1 + q.max(-1).min(1),
            );
        }
    }

    fn solve(mut self) -> usize {
        for Motion { direction, count } in self.motions {
            for _ in 0..*count {
                self.knots[0] = direction.step_signed(self.knots[0]);
                (1..self.knots.len()).for_each(|i| self.follow(i));
                self.visited.insert(*self.knots.last().unwrap());
            }
        }
        self.visited.len()
    }
}

fn setup(input: &str) -> Input {
    input
        .trim()
        .lines()
        .map(|line| Motion {
            direction: line.chars().next().unwrap().into(),
            count: line[2..].parse().unwrap(),
        })
        .collect()
}

fn part1(input: &Input) -> usize {
    Solver::new(input, 2).solve()
}

fn part2(input: &Input) -> usize {
    Solver::new(input, 10).solve()
}

aoc::main!(2022, 9);
aoc::example!(ex01, "09.1.txt", 13, 1);
aoc::example!(ex02, "09.2.txt", 88, 36);
aoc::test_input!("09.txt", 6030, 2545);
